### DISCUSSION TITLE
Streaming without "with" like urllib3

### DISCUSSION THREAD

**Discussion by GunGunGun** [encode/httpcore#650] [General]
In urllib3, I can use r = pool.urlopen('GET', url, preload_content=False) to stream content, and it's very useful because I can then use r.headers to get only response headers, then use r.data to get content. It seems very hard to do so in httpcore because I need to use with to wrap httpcore.stream('GET', url), but I find it's very limited, for example I have a local transparent web filter proxy, it's better to be able to fetch only response headers, filter/change response headers first, then download content later if response headers is not blocked.

In urllib3 I use:

```
import urllib3
p = urllib3.PoolManager()
r = p.urlopen('https://example.com', preload_content=False)
h = r.headers
# Filter h > headers, for example:
if 'google.com' in h.get('Referer'):
    # Abort request
# If h is okay then
d = r.data
# Then send to client
```

**Comment by lovelydinosaur** [comment]
So... the `with` block forces the opening/closing context to be explicit. 

You should be able to get the same behaviour, but it'll ensure you've got good guardrails and can't accidentally leave the response in an open state...

```python
import httpcore


p = httpcore.ConnectionPool()

with p.stream('GET', 'https://example.com') as r:
    h = dict(r.headers)
    if 'google.com' in h.get('Referer', ''):
        # Abort request

    # If h is okay then
    r.read()
    d = r.content
    # Then send to client
```
